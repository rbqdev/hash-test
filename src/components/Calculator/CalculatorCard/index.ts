export { default } from "./CalculatorCard";
export type { ICalculatorCard } from "./CalculatorCard";
