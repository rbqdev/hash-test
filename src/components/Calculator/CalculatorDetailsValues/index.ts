export { default } from "./CalculatorDetailsValues";
export type { ICalculatorCardValues } from "./CalculatorDetailsValues";
