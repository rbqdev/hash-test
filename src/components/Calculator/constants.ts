export const NAME_AMMOUNT = "ammount";
export const NAME_INSTALLMENTS = "installments";
export const NAME_MDR = "mdr";
export const DEBOUNCE_TIME_FOR_TRIGGER_REQUEST = 500;
