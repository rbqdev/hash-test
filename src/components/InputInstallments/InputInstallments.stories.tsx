import { Story, Meta } from "@storybook/react/types-6-0";
import InputInstallments from ".";
import { ICustomInput } from "@components/InputText";

export default {
  title: "InputInstallments",
  component: InputInstallments,
  args: {
    error: false
  }
} as Meta;

export const Default: Story<ICustomInput> = args => (
  <div style={{ maxWidth: 300, padding: 15 }}>
    <InputInstallments {...args} />
  </div>
);
