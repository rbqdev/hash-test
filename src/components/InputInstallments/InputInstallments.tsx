import { useRef, memo } from "react";
import InputText, { ICustomInput } from "@components/InputText";
import getOnlyDigits from "@utils/getOnlyDigits";

export const LABEL_INSTALLMENTS = "Em quantas parcelas *";

function InputInstallments({ callbackFunction, name, error }: ICustomInput) {
  const inputRef = useRef<HTMLInputElement>(null);

  const handleChange = (): void => {
    const element = inputRef.current;

    if (element === null) {
      return;
    }

    element.value = `${getOnlyDigits(element.value)}`;

    if (typeof callbackFunction !== "function") {
      return;
    }

    callbackFunction({
      value: getOnlyDigits(element.value),
      name
    });
  };

  return (
    <InputText
      label={LABEL_INSTALLMENTS}
      aria-label={LABEL_INSTALLMENTS}
      aria-required="true"
      textInfo="MÃ¡ximo de 12 parcelas"
      name={name}
      error={error}
      onChange={handleChange}
      ref={inputRef}
    />
  );
}

export default memo(InputInstallments);
