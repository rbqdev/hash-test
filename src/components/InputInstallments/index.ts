export { default, LABEL_INSTALLMENTS } from "./InputInstallments";
