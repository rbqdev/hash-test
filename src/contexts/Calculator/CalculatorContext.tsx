import React, { createContext } from "react";
import useFetch, { IInitialState } from "@hooks/useFetch/useFetch";
import getUrlAnticipation from "@utils/getUrlAnticipation";

interface IRequestBody {
  amount: number;
  installments: number;
  mdr: number;
  days: number[];
}

interface ICalculatorContext extends IInitialState {
  requestAnticipation: (payload: IRequestBody) => void;
  resetFetchValues: () => void;
}

const CalculatorContext = createContext({} as ICalculatorContext);

export const CalculatorProvider: React.FunctionComponent = ({ children }) => {
  const {
    response,
    error,
    isLoading,
    requestTakeToLong,
    dispatchRequest,
    resetFetchValues
  } = useFetch();

  const requestAnticipation = async (payload: IRequestBody): Promise<void> => {
    const url = getUrlAnticipation();

    await dispatchRequest({
      url,
      method: "POST",
      body: payload
    });
  };

  return (
    <CalculatorContext.Provider
      value={{
        response,
        error,
        isLoading,
        requestTakeToLong,
        requestAnticipation,
        resetFetchValues
      }}
    >
      {children}
    </CalculatorContext.Provider>
  );
};

export default CalculatorContext;
