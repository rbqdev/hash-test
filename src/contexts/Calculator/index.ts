export { default, CalculatorProvider } from "./CalculatorContext";
