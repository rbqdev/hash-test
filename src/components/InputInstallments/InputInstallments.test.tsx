import render from "@testsUtils/renderWithTheme";
import { NAME_INSTALLMENTS } from "@components/Calculator/constants";
import { RenderResult } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import InputInstallments from ".";

interface PropsRender {
  callbackFunction: jest.Mock<any, any> | undefined;
  error: boolean;
}

const mockCallbackFunction = jest.fn();

const renderComponent = ({ callbackFunction, error }: PropsRender) =>
  render(
    <InputInstallments
      name={NAME_INSTALLMENTS}
      callbackFunction={callbackFunction}
      error={error}
    />
  );

describe("Component: InputInstallments", () => {
  const inputRole = { name: /em quantas parcelas /i };
  let inputInstallments: HTMLElement;

  it("should render component correctly", () => {
    const { baseElement } = renderComponent({
      error: false,
      callbackFunction: mockCallbackFunction
    });

    expect(baseElement).toMatchSnapshot();
  });

  describe("when 'callbackFunction' is passed", () => {
    let wrapper: RenderResult;

    beforeEach(() => {
      wrapper = renderComponent({
        error: false,
        callbackFunction: mockCallbackFunction
      });

      inputInstallments = wrapper.getByRole("textbox", inputRole);

      userEvent.type(inputInstallments, "12");
    });

    it("should set value", () => {
      expect(inputInstallments).toHaveValue("12");
    });

    it("should render text info", () => {
      expect(wrapper.queryByText("MÃ¡ximo de 12 parcelas")).toBeInTheDocument();
    });

    it("should call 'callbackFunction' when user type on input", () => {
      inputInstallments = wrapper.getByRole("textbox", inputRole);

      userEvent.type(inputInstallments, "12");

      expect(mockCallbackFunction).toBeCalledWith({
        value: 12,
        name: NAME_INSTALLMENTS
      });
    });
  });

  describe("when 'callbackFunction' not passed", () => {
    it("should not be called when user type on input", () => {
      const { getByRole } = renderComponent({
        error: false,
        callbackFunction: undefined
      });

      inputInstallments = getByRole("textbox", inputRole);

      userEvent.type(inputInstallments, "12");

      expect(mockCallbackFunction).not.toHaveBeenCalled();
    });
  });
});
