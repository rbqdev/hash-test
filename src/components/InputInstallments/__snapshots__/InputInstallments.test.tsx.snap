// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: InputInstallments should render component correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  margin-bottom: 22px;
}

.c0:last-child {
  margin-bottom: 0;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c2 {
  color: #656565;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  font-size: 14px;
  line-height: 18px;
  margin-bottom: 6px;
}

.c5 {
  color: #cecece;
  font-size: 11px;
  line-height: 14px;
  font-weight: bold;
  margin-top: 4px;
}

.c3 {
  position: relative;
}

.c4 {
  color: #656565;
  background: #fff;
  border: 1px solid #d1dce3;
  border-radius: 4px;
  box-sizing: border-box;
  font-size: 14px;
  line-height: 18px;
  padding: 9px 28px 9px 12px;
  width: 100%;
  outline: none;
}

.c4:hover {
  border: 1px solid #b1c4d0;
}

.c4:focus {
  border: 1px solid #66afe9;
}

<body>
  <div>
    <div
      class="c0"
    >
      <label
        class="c1"
        for="installments"
      >
        <span
          class="c2"
        >
          Em quantas parcelas *
        </span>
        <div
          class="c3"
        >
          <input
            aria-invalid="false"
            aria-label="Em quantas parcelas *"
            aria-required="true"
            class="c4"
            id="installments"
            name="installments"
            type="text"
          />
        </div>
        <small
          class="c5"
        >
          MÃ¡ximo de 12 parcelas
        </small>
      </label>
    </div>
  </div>
</body>
`;
